README.md
# NLP
Tagging, parsing jobs.
HMMs, Viterbi, Log-linear Models


h2-p/history
   27  vim matrix_binary_writer(2).cpp 
   28  vim matrix_binary_writer\(2\).cpp 
   29  gcc matrix_binary_writer.cpp 
   30  gcc matrix_binary_writer\(2\).cpp 
   31  ./a.out 
   32  vim matrix.dat 
   33  cat matrix.dat
   34  ./a.out 
   35  cat matrix.dat
   36  vim matrix_binary_writer\(2\).cpp 
   37  vim matrix.dat
   38  gcc matrix_binary_writer\(2\).cpp 
   39  ./a.out 
   40  cat matrix.dat 
   41  vim matrix_binary_writer\(2\).cpp 
   42  gcc matrix_binary_writer\(2\).cpp 
   43  vim matrix_binary_writer\(2\).cpp 
   44  gcc matrix_binary_writer\(2\).cpp 
   45  vim matrix_binary_writer\(2\).cpp 
   46  clear
   47  vim matrix_binary_writer.cpp 
   48  gcc matrix_binary_writer.cpp 
   49  vim matrix_binary_writer.cpp 
   50  gcc matrix_binary_writer.cpp 
   51  gcc matrix_binary_writer.cpp 
   52  vim matrix_binary_writer.cpp 
   53  vim matrix_binary_writer\(2\).cpp 
   54  vim matrix_binary_writer.cpp 
   55  vim matrix_binary_writer\(2\).cpp 
   56  vim matrix_binary_writer.cpp 
   57  gcc matrix_binary_writer.cpp 
   58  vim matrix_binary_writer.cpp 
   59  vim matrix_binary_writer.cpp 
   60  gcc matrix_binary_writer.cpp 
   61  vim matrix_binary_writer.cpp 
   62  gcc matrix_binary_writer.cpp 
   63  vim matrix_binary_writer\(2\).cpp 
   64  gcc matrix_binary_writer\(2\).cpp 
   65  vim matrix_binary_writer\(2\).cpp 
   66  gcc matrix_binary_writer\(2\).cpp 
   67  vim matrix_binary_writer\(2\).cpp 
   68  gcc matrix_binary_writer\(2\).cpp 
   69  vim matrix_binary_writer\(2\).cpp 
   70  history
   71  ls -l
   72  python main.py 
   73  vim main.py 
   74  python main.py 
   75  vim main.py 
   76  python main.py 
   77  vim main.py 
   78  vim main.py 
   79  python main.py 
   80  vim main.py 
   81  python main.py 
   82  vim main.py 
   83  python main.py 
   84  vim main.py 
   85  vim main.py 
   86  python main.py 
   87  vim main.py 
   88  python main.py 
   89  vim main.py 
   90  python main.py 
   91  python main.py 
   92  python main.py 
   93  python main.py 
   94  python main.py 
   95  python main.py 
   96  python main.py 
   97  ls -l
   98  jupyter notebook
   99  python main.py 
  100  python main.py 
  101  history
  102  history
  103  python main.py 
  104  python
  105  ls -l
  106  jupyter notebook
  107  ls -l
  108  pwd
  109  cd ..
  110  cd..
  111  ls -l
  112  cd ..
  113  ls -l
  114  cd ..
  115  ls -l
  116  cd deeplearning/
  117  ls -l
  118  cd RNNs/
  119  ls
  120  jupyter notebook
  121  ls -l
  122  cd ..
  123  ls
  124  cd ..
  125  cd Projects/VanGogh/
  126  vim mai
  127  cd neuralart_tensorflow/
  128  vim main.py 
  129  python
  130  vim main.py 
  131  vim main.py 
  132  vim main.py 
  133  vim main.py 
  134  vim main.py 
  135  vim main.py 
  136  vim main.py 
  137  vim main.py 
  138  vim main.py 
  139  vim main.py 
  140  vim main.py 
  141  vim main.py 
  142  vim main.py 
  143  python main.py 
  144  vim main.py 
  145  ssh wuhao@dhcp-c101a5c6.ucd.ie
  146  ssh wuhao@dhcp-c101a5c6.ucd.ie
  147  vim main.py 
  148  vim main.py 
  149  python main.py 
  150  vim main.py 
  151  python main.py 
  152  vim main.py 
  153  python main.py 
  154  vim main.py 
  155  python main.py 
  156  vim main.py 
  157  python main.py 
  158  vim main.py 
  159  vim main.py 
  160  vim main.py 
  161  python main.py 
  162  vim main.py 
  163  python main.py 
  164  python main.py 
  165  vim main.py 
  166  python main.py 
  167  vim main.py 
  168  python main.py 
  169  cd ~/Desktop/blog/
  170  ls
  171  ssh -i "GraphPatternMatching.pem" root@ec2-54-217-197-4.eu-west-1.compute.amazonaws.com
  172  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-217-197-4.eu-west-1.compute.amazonaws.com
  173  chmod 400 GraphPatternMatching.pem
  174  chmod 400 GraphPatternMatching.pem
  175  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-217-197-4.eu-west-1.compute.amazonaws.com
  176  scp -i "GraphPatternMatching.pem" eng.tar.gz ubuntu@54.155.191.173:/home/ubuntu/download/GroundedTranslation-master/data
  177  scp -i "GraphPatternMatching.pem" ger.tar.gz ubuntu@54.155.191.173:/home/ubuntu/download/GroundedTranslation-master/data/ger
  178  pwd
  179  ssh -i "GraphPatternMatching.pem" root@ec2-54-78-250-26.eu-west-1.compute.amazonaws.com
  180  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-78-250-26.eu-west-1.compute.amazonaws.com
  181  ssh -i "GraphPatternMatching.pem" root@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  182  ssh -i "GraphPatternMatching.pem" ubunt@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  183  pwd
  184  ssh -i "GraphPatternMatching.pem" ubunt@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  185  ssh -i "GraphPatternMatching.pem" ubunt@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  186  ls
  187  lsssh -i "GraphPatternMatching.pem" root@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  188  ssh -i "GraphPatternMatching.pem" root@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  189  ssh -i "GraphPatternMatching.pem" root@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  190  ssh -i "GraphPatternMatching.pem" root@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  191  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-155-103-63.eu-west-1.compute.amazonaws.com
  192  ssh -i "GraphPatternMatching.pem" root@ec2-54-170-184-72.eu-west-1.compute.amazonaws.com
  193  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-170-184-72.eu-west-1.compute.amazonaws.com
  194  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-216-141-164.eu-west-1.compute.amazonaws.com
  195  cp ~/Downloads/GraphPatternMatching.pem ~/Desktop/blog/
  196  cd ~/Desktop/blog/
  197  ls -l
  198  sudo vim ~/.bash_profile 
  199  ssh -i "GraphPatternMatching.pem" root@ec2-54-217-146-208.eu-west-1.compute.amazonaws.com
  200  ssh -i "GraphPatternMatching.pem" ubuntu@ec2-54-217-146-208.eu-west-1.compute.amazonaws.com
  201  cd ~/Documents/DITstudy/Coursera/PGMs/Codes/
  202  jupyter notebook
  203  pwd
  204  ls
  205  ls
  206  pwd

h1-p/q1.py

# coding: utf-8

# In[11]:

#! /usr/bin/python

__author__="Daniel Bauer <bauer@cs.columbia.edu>"
__date__ ="$Sep 12, 2011"

import sys
from collections import defaultdict
from operator import itemgetter
import math

"""
Count n-gram frequencies in a data file and write counts to
stdout. 
"""

def simple_conll_corpus_iterator(corpus_file):
    """
    Get an iterator object over the corpus file. The elements of the
    iterator contain (word, ne_tag) tuples. Blank lines, indicating
    sentence boundaries return (None, None).
    """
    l = corpus_file.readline()
    while l:
        line = l.strip()
        if line: # Nonempty line
            # Extract information from line.
            # Each line has the format
            # word pos_tag phrase_tag ne_tag
            fields = line.split(" ")
            ne_tag = fields[-1] # tag
            #phrase_tag = fields[-2] #Unused
            #pos_tag = fields[-3] #Unused
            word = " ".join(fields[:-1])
            yield word, ne_tag
        else: # Empty line
            yield (None, None)                        
        l = corpus_file.readline()

def sentence_iterator(corpus_iterator):
    """
    Return an iterator object that yields one sentence at a time.
    Sentences are represented as lists of (word, ne_tag) tuples.
    """
    current_sentence = [] #Buffer for the current sentence
    for l in corpus_iterator:    
#             print ("***NOW",l) # Every line (every word) l -->  ('nucleotidase', 'I-GENE')
            if l==(None, None):
                if current_sentence:  #Reached the end of a sentence
                    yield current_sentence
                    current_sentence = [] #Reset buffer
                else: # Got empty input stream
                    sys.stderr.write("WARNING: Got empty input file/stream.\n")
                    raise StopIteration
            else:
                current_sentence.append(l) #Add token to the buffer

    if current_sentence: # If the last line was blank, we're done
#         print ("Current ",current_sentence)
        yield current_sentence  #Otherwise when there is no more token
                                # in the stream return the last sentence.
        #Current_sentence  [('When', 'O'), ('CSF', 'O'), ('joj', 'O'),...]  (read in as a key-value pairs collection)

def get_ngrams(sent_iterator, n):
    """
    Get a generator that returns n-grams over the entire corpus,
    respecting sentence boundaries and inserting boundary tokens.
    Sent_iterator is a generator object whose elements are lists
    of tokens. 每个句子前追加2个*，后追加一个STOP标志(w_boundary)
    """
    for sent in sent_iterator:
         # sent (list) --> [('omparison', 'O'), ('with', 'O'),  ...] (Tupels in every sentence)
         #Add boundary symbols to the sentence
         w_boundary = (n-1) * [(None, "*")]
         w_boundary.extend(sent)
         w_boundary.append((None, "STOP"))
         #Then extract n-grams
        
         ngrams = (tuple(w_boundary[i:i+n]) for i in range(len(w_boundary)-n+1))
         for n_gram in ngrams: #Return one n-gram at a time
#             print (n_gram)
            yield n_gram      

#         ngrams
#         (1, 2, 3)
#         (2, 3, 4)
#         (3, 4, 5)
#         (4, 5, 6)
#         (5, 6, 7)
#         >>> 类似这种
#  返回的是这样：
# ((None, '*'), (None, '*'), ('omparison', 'O'))
# ((None, '*'), ('omparison', 'O'), ('with', 'O'))
# (('omparison', 'O'), ('with', 'O'), ('alkaline', 'I-GENE'))


class Hmm(object):
    """
    Stores counts for n-grams and emissions. 
    """

    def __init__(self, n=3):
        assert n>=2, "Expecting n>=2."
        self.n = n
        self.emission_counts = defaultdict(int)
        self.ngram_counts = [defaultdict(int) for i in range(self.n)]
        self.all_states = set()
#       stores emmision para: same type with counts. shape (word,tag):value   meaning em (word|tag)
        self.emision_parameters = defaultdict(float)
#       rare word list
        self.rare = list()
#       argmax word list :    word([tag,em],[tag,em],...)
        self.argmax_list = defaultdict(list)
        self.argmax_value = defaultdict(str)


    def train(self, corpus_file):
        """
        Count n-gram frequencies and emission probabilities from a corpus file.
        """
        ngram_iterator =             get_ngrams(sentence_iterator(simple_conll_corpus_iterator(corpus_file)), self.n)
        

        for ngram in ngram_iterator:
            #Sanity check: n-gram we get from the corpus stream needs to have the right length
            assert len(ngram) == self.n, "ngram in stream is %i, expected %i" % (len(ngram, self.n))

            tagsonly = tuple([ne_tag for word, ne_tag in ngram]) #retrieve only the tags    
        #  tagsonly 只收集 ne_tag
        #     ('*', '*', 'O')
        #     ('*', 'O', 'O')
        #     ('O', 'O', 'I-GENE')
            
            for i in range(2, self.n+1): #Count NE-tag 2-grams..n-grams
                self.ngram_counts[i-1][tagsonly[-i:]] += 1
            if ngram[-1][0] is not None: # If this is not the last word in a sentence
                self.ngram_counts[0][tagsonly[-1:]] += 1 # count 1-gram
                self.emission_counts[ngram[-1]] += 1 # and emission frequencies

            # Need to count a single n-1-gram of sentence start symbols per sentence
            if ngram[-2][0] is None: # this is the first n-gram in a sentence
                self.ngram_counts[self.n - 2][tuple((self.n - 1) * ["*"])] += 1
#             print (self.emission_counts)


    def write_counts(self, output, printngrams=[1,2,3]):
        """
        Writes counts to the output file object.
        Format:

        """
        # First write counts for emissions
        for word, ne_tag in self.emission_counts:            
            output.write("%i WORDTAG %s %s\n" % (self.emission_counts[(word, ne_tag)], ne_tag, word))


        # Then write counts for all ngrams
        for n in printngrams:            
            for ngram in self.ngram_counts[n-1]:
                ngramstr = " ".join(ngram)
                output.write("%i %i-GRAM %s\n" %(self.ngram_counts[n-1][ngram], n, ngramstr))

    def read_counts(self, corpusfile):

        self.n = 3
        self.emission_counts = defaultdict(int)
        self.ngram_counts = [defaultdict(int) for i in range(self.n)]
        self.all_states = set()

        for line in corpusfile:
            parts = line.strip().split(" ")
            count = float(parts[0])
            if parts[1] == "WORDTAG":
                ne_tag = parts[2]
                word = parts[3]
                self.emission_counts[(word, ne_tag)] = count
                self.all_states.add(ne_tag)
            elif parts[1].endswith("GRAM"):
                n = int(parts[1].replace("-GRAM",""))
                ngram = tuple(parts[2:])
                self.ngram_counts[n-1][ngram] = count
        
    
    def get_emission_paras(self):
    
        self.emision_parameters = defaultdict(float)
        for wordNtag,counts in self.emission_counts.items():
            tag = wordNtag[1] #get tag of the tuple
            tagCount = self.ngram_counts[0][(tag,)]
            emission = counts / tagCount
            self.emision_parameters[wordNtag]=emission
#             print (wordNtag,emission)
#         print (self.emision_parameters)
            
    def replaceRare(self,training_file,trainingChange_file):
        for wordNtag,counts in self.emission_counts.items():
            if counts < 5:
                print ('Now',wordNtag,counts)
                self.rare.append(wordNtag[0])
                
#         print (self.rare)
        
        l = training_file.readline()
        print (l)
        i = 0
        while l:
            if i%1000==0: print ('Now line',i)
	    i=i+1
            line = l.strip()
            if line: # Nonempty line
                
                fields = line.split(" ")
                ne_tag = fields[-1] # tag
                word = fields[0]
#                 word = " ".join(fields[:-1])
                if word in self.rare:
#                     print ('found a word',word)
                    newline="%s %s\n" % ('_RARE_',ne_tag)
                    trainingChange_file.write(newline)
                else:
                    trainingChange_file.write(l)

            else: # Empty line
                l = training_file.readline()
                trainingChange_file.write('\n')

#                 print ("Not Replaced!!!")
                
            l = training_file.readline()
        
        trainingChange_file.close()


    def get_argmax(self):
#         buid argmax list
        for wordNtag,values in self.emision_parameters.items():
#             if wordNtag[0]=='_RARE_':
                
            self.argmax_list[wordNtag[0]].append((wordNtag[1],values))
#         build argmax real tag
        for word,values in self.argmax_list.items():
            tag = max(values,key=itemgetter(1))[0]
            self.argmax_value[word]=tag
            
        print(self.argmax_value)
        
    def write_prediction(self,testing,output):
        # read in word first
        for line in testing:
            word = line.strip()
            if word:
                tag=self.argmax_value[word]
                if tag:
                    output.write("%s %s\n" % (word, tag))
                else:
                    tag=self.argmax_value['_RARE_']
#                     print ('New word',tag)
                    output.write("%s %s\n" % (word, tag))
            else:
                output.write("\n")
        
                
#         # write (word tag) 
#         for word, tag in self.argmax_value.items():            
#             output.write("%s %s\n" % (word, tag))
        print ("Finish!")
        output.close()

        


def usage():
    print ("""
    python count_freqs.py [input_file] > [output_file]
        Read in a gene tagged training input file and produce counts.
    """)

if __name__ == "__main__":

    try:
#         input = file(sys.argv[1],"r")
        input = open("gene.counts",'r')
        training = open("gene.train",'r')
        trainingC = open("gene.train.replaced",'w')
        
        testing = open('gene.dev','r')
        prediction = open("gene.counts.out",'w');
        
    except IOError:
        sys.stderr.write("ERROR: Cannot read inputfile %s.\n" % arg)
        sys.exit(1)
    
    # Initialize a trigram counter
    counter = Hmm(3)
    # Collect counts
#     counter.train(input)
    
    #Q1a: Read the counts
    counter.read_counts(input)
    counter.get_emission_paras()
    
    #Q1b: Replace
    counter.replaceRare(training,trainingC)

#     Q1c: Argmax
#     counter.get_argmax()
#     counter.write_prediction(testing,prediction)


# In[ ]:




# In[ ]:


# 47
# down vote
# accepted
# Use max():

 
# Using itemgetter():

# In [53]: lis=[(101, 153), (255, 827), (361, 961)]

# In [81]: from operator import itemgetter

# In [82]: max(lis,key=itemgetter(1))[0]    #faster solution
# Out[82]: 361

